File: refinery-forcefield/src/main/native/cpp/LineCharge.cpp
#include "refinery/forcefield/LineCharge.h"

namespace refinery::forcefield {

frc::Translation2d LineCharge::Evaluate(const frc::Translation2d& point) const {
  frc::Translation2d segmentVec = m_end - m_start;
  double segLenSq = segmentVec.Norm().value() * segmentVec.Norm().value();

  // Degenerate segment — treat as a point
  if (segLenSq < 1e-9) {
    frc::Translation2d delta = m_start - point;
    double r = delta.Norm().value();
    if (r < kMinDistance) r = kMinDistance;
    double magnitude = m_strength * std::max(0.0, 1.0 - r / m_falloffDistance);
    if (r < 1e-9) return frc::Translation2d{};
    return (delta / delta.Norm().value()) * magnitude;
  }

  // Project point onto segment
  frc::Translation2d startToPoint = point - m_start;
  double t = (startToPoint.X().value() * segmentVec.X().value() +
              startToPoint.Y().value() * segmentVec.Y().value()) /
             segLenSq;
  t = std::clamp(t, 0.0, 1.0);

  frc::Translation2d closest = m_start + segmentVec * t;
  frc::Translation2d delta = closest - point;
  double r = delta.Norm().value();

  if (r > m_falloffDistance) {
    return frc::Translation2d{};
  }
  if (r < kMinDistance) {
    r = kMinDistance;
  }

  double magnitude = m_strength * (1.0 - r / m_falloffDistance);

  if (delta.Norm().value() < 1e-9) {
    return frc::Translation2d{};
  }

  return (delta / delta.Norm().value()) * magnitude;
}

}  // namespace refinery::forcefield

File: refinery-forcefield/src/main/java/com/bionanomics/refinery/forcefield/LineCharge.java
package com.bionanomics.refinery.forcefield;

import edu.wpi.first.math.geometry.Translation2d;

/**
 * A line charge that creates force perpendicular to a wall or barrier segment.
 *
 * <p>The force is computed by projecting the query point onto the line segment and applying
 * a force perpendicular to the segment. The force falls off linearly from full strength at
 * the segment surface to zero at {@code falloffDistance} meters away.
 *
 * <p>Positive strength = attractive (pulls toward the line), negative = repulsive (pushes away).
 * Walls should use negative strength to repel.
 */
public class LineCharge implements Charge {
    private static final double kMinDistance = 0.02;

    private final String id;
    private final Translation2d start;
    private final Translation2d end;
    private final double strength;
    private final double falloffDistance;

    /**
     * @param id              Unique identifier
     * @param start           Start point of the line segment in field meters
     * @param end             End point of the line segment in field meters
     * @param strength        Force magnitude; positive = attractive, negative = repulsive
     * @param falloffDistance Distance in meters at which force reaches zero
     */
    public LineCharge(String id, Translation2d start, Translation2d end,
                      double strength, double falloffDistance) {
        this.id = id;
        this.start = start;
        this.end = end;
        this.strength = strength;
        this.falloffDistance = Math.max(falloffDistance, kMinDistance);
    }

    @Override
    public Translation2d evaluate(Translation2d point) {
        Translation2d segmentVec = end.minus(start);
        double segmentLengthSquared = segmentVec.getNorm() * segmentVec.getNorm();

        // Degenerate segment (zero length) — treat as a point
        if (segmentLengthSquared < 1e-9) {
            Translation2d delta = start.minus(point);
            double r = delta.getNorm();
            if (r < kMinDistance) r = kMinDistance;
            double magnitude = strength * Math.max(0, 1.0 - r / falloffDistance);
            if (r < 1e-9) return Translation2d.kZero;
            return delta.div(delta.getNorm()).times(magnitude);
        }

        // Parameter t along the segment: t=0 at start, t=1 at end
        Translation2d startToPoint = point.minus(start);
        double t = (startToPoint.getX() * segmentVec.getX()
                  + startToPoint.getY() * segmentVec.getY()) / segmentLengthSquared;
        t = Math.max(0, Math.min(1, t));

        // Closest point on segment
        Translation2d closest = start.plus(segmentVec.times(t));

        // Vector from point toward the closest point on the segment
        Translation2d delta = closest.minus(point);
        double r = delta.getNorm();

        if (r > falloffDistance) {
            return Translation2d.kZero;
        }
        if (r < kMinDistance) {
            r = kMinDistance;
        }

        double magnitude = strength * (1.0 - r / falloffDistance);

        if (delta.getNorm() < 1e-9) {
            return Translation2d.kZero;
        }

        Translation2d direction = delta.div(delta.getNorm());
        return direction.times(magnitude);
    }

    @Override public String id() { return id; }
    @Override public double strength() { return strength; }
    public Translation2d start() { return start; }
    public Translation2d end() { return end; }
    public double falloffDistance() { return falloffDistance; }
}
